test/internal/pbxproj_partials/write_pbxproject_targets_tests.bzl

```diff
@@ -0,0 +1,457 @@
+_TARGET_DEPENDENCIES_DECLARED_FILE = "a_generator_name_pbxproj_partials/pbxtargetdependencies"
+
+_PLATFORM_NAME = {
+    "IOS_DEVICE": "iphoneos",
+    "IOS_SIMULATOR": "iphonesimulator",
+    "MACOS": "macosx",
+    "TVOS_DEVICE": "appletvos",
+    "TVOS_SIMULATOR": "appletvsimulator",
+    "WATCHOS_DEVICE": "watchos",
+    "WATCHOS_SIMULATOR": "watchsimulator",
+}
+
+def _apple_platform_to_platform_name(platform):
+    return _PLATFORM_NAME[platform]
+
+def _consolidation_map_declared_file(idx):
+    return "a_generator_name_pbxproj_partials/consolidation_maps/{}".format(idx)
+
+def _quote_if_needed(str):
+    if " " in str:
+        return "'{}'".format(str)
+    return str
+
+def _json_to_xcode_targets_by_label(json_str):
+    return {
+        Label(label): _dict_to_xcode_targets_by_id(d)
+        for label, d in json.decode(json_str).items()
+    }
+
+def _dict_to_xcode_targets_by_id(d):
+    return {
+        id: _dict_to_xcode_target(xcode_target_dict)
+        for id, xcode_target_dict in d.items()
+    }
+
+def _dict_to_xcode_target(d):
+    return struct(
+        id = d["id"],
+        dependencies = depset(d["dependencies"]),
+        platform = struct(
+            arch = d["arch"],
+            os_version = d["os_version"],
+            platform = d["platform"],
+        ),
+        product = struct(
+            path = d["product_path"],
+            type = d["product_type"],
+        ),
+        test_host = d["test_host"],
+    )
+
+# FIXME: Extract
+def mock_xcode_target(
+    *,
+    id,
+    arch,
+    dependencies,
+    os_version,
+    platform,
+    product_path,
+    product_type,
+    test_host = None):
+    return struct(
+        id = id,
+        arch = arch,
+        dependencies = dependencies,
+        os_version = os_version,
+        platform = platform,
+        product_path = product_path,
+        product_type = product_type,
+        test_host = test_host,
+    )
+
+def _write_pbxproject_targets_test_impl(ctx):
+    env = unittest.begin(ctx)
+
+    # Arrange
+
+    args = []
+
+    def _args_add_all(flag_or_values, values = None, *, map_each = None):
+        if values != None:
+            flag = flag_or_values
+        else:
+            flag = None
+            values = flag_or_values
+
+        if type(values) == "depset":
+            values = values.to_list()
+
+        if not values:
+            return
+
+        if flag:
+            args.append(flag)
+
+        if map_each:
+            for value in values:
+                mapped_value = map_each(value)
+                if type(mapped_value) == "list":
+                    args.extend(
+                        [_quote_if_needed(str(v)) for v in mapped_value],
+                    )
+                elif mapped_value:
+                    args.append(_quote_if_needed(str(mapped_value)))
+        else:
+            args.extend([_quote_if_needed(str(value)) for value in values])
```

Comment:
After this PR is merged, I'll open another consolidating/cleaning up these helper functions, now that we have 3 related implementations.